### PR TITLE
GC coverage can now occur in scenarios where a thread doesn't exist.

### LINKED ISSUE #33612: Test failure: Interop/NativeCallable/NativeCallableTest/NativeCallableTest.sh
**Job:**
[runtime-coreclr gcstress0x3-gcstress0xc:20200315.1](https://dev.azure.com/dnceng/public/_build/results?buildId=560162)
**Error:**
```
Assert failure(PID 202 [0x000000ca], Thread: 210 [0x00d2]): pThread
File: /__w/1/s/src/coreclr/src/vm/gccover.cpp Line: 1403
Image: /root/helix/work/correlation/corerun\n\n/root/helix/work/workitem/Interop/NativeCallable/NativeCallableTest/NativeCallableTest.sh: line 275: 202 Aborted $LAUNCHER $ExePath "${CLRTestExecutionArguments[@]}"

Return code: 1
Raw output file: /root/helix/work/workitem/Interop/NativeCallable/Reports/Interop.NativeCallable/NativeCallableTest/NativeCallableTest.output.txt
Raw output:
BEGIN EXECUTION\n/root/helix/work/correlation/corerun NativeCallableTest.dll ''
Running TestNativeCallableValid...
Running TestNativeCallableValid_OnNewNativeThread...
Running TestNativeCallableValid_PrepareMethod...
Writing minidump with heap to file /home/helixbot/dotnetbuild/dumps/coredump.202.dmp
Written 60399616 bytes (14746 pages) to core file
Expected: 100
Actual: 134
END EXECUTION - FAILED
Test Harness Exitcode is : 1
To run the test:
> set CORE_ROOT=/root/helix/work/correlation
> /root/helix/work/workitem/Interop/NativeCallable/NativeCallableTest/NativeCallableTest.sh
Expected: True
Actual: False
```
**Stack trace:**
`
at Interop_NativeCallable._NativeCallableTest_NativeCallableTest_._NativeCallableTest_NativeCallableTest_sh() in /__w/1/s/artifacts/tests/coreclr/Linux.arm.Checked/TestWrappers/Interop.NativeCallable/Interop.NativeCallable.XUnitWrapper.cs:line 140
`
**Details:**
https://dev.azure.com/dnceng/public/_build/results?buildId=560162&view=ms.vss-test-web.build-test-results-tab&runId=17644952&resultId=101937&paneView=debug

### PR DESCRIPTION
**Pull request by AaronRobinsonMSFT**
Properly handle this case and assert some invariant.

Fixes #33612 

/cc @jkotas

### CODE REVIEW

**Review by jkotas — APPROVED** [review]

File: src/coreclr/src/vm/gccover.cpp [hunk 1]
```diff
@@ -1400,7 +1400,15 @@ BOOL OnGcCoverageInterrupt(PCONTEXT regs)
...
     Thread* pThread = GetThread();
-    _ASSERTE(pThread);
+    if (!pThread)
+    {
+        // No thread at the moment so we aren't doing coverage for this function.
+        // This should only occur for methods with the NativeCallableAttribute,
+        // where the caller could be coming from a thread unknown to the CLR and
+        // we haven't created a thread yet - see PreStubWorker_Premptive().
+        _ASSERTE(pMD->HasNativeCallableAttribute());
+        return (FALSE);
```
Comment (jkotas) [inline src/coreclr/src/vm/gccover.cpp]: Does returning FALSE here make the test to work?

I expect we are going to hit `CONSISTENCY_CHECK_MSG(!ExecutionManager::IsManagedCode(GetIP(pContext)), "hit privileged instruction!");` right after the `OnGcCoverageInterrupt` call.

I think this needs to do `RemoveGcCoverageInterrupt` and `return TRUE;`.

Comment (AaronRobinsonMSFT) [reply src/coreclr/src/vm/gccover.cpp]: Hmmm. Let me see what I did wrong then. I see what you are saying.

Comment (AaronRobinsonMSFT) [reply src/coreclr/src/vm/gccover.cpp]: You're right. I was running the test under RET not CHK. I see the error now.

File: src/coreclr/src/vm/jitinterface.cpp [hunk 2]
```diff
@@ -12678,9 +12678,10 @@ PCODE UnsafeJitFunction(PrepareCodeConfig* config,
 
     flags = GetCompileFlags(ftn, flags, &methodInfo);
 
-    // If the reverse P/Invoke flag is used, we aren't going to support
-    // any tiered compilation.
-    if (flags.IsSet(CORJIT_FLAGS::CORJIT_FLAG_REVERSE_PINVOKE))
+    // Clearing all tier flags and mark as optimized if the reverse P/Invoke
```
Comment (jkotas) [inline src/coreclr/src/vm/jitinterface.cpp]: Nit: Should this whole `if` be under `#ifdef FEATURE_TIERED_COMPILATION`?

**Review by jkotas — APPROVED** [review]

File: src/coreclr/src/vm/jitinterface.cpp (hunk 2 above)
Comment (AaronRobinsonMSFT) [reply src/coreclr/src/vm/jitinterface.cpp]: We could. The `IsEligibleForTieredCompilation()` isn't under that so I didn't think it was needed. The `IsEligibleForTieredCompilation` is hard coded outside the `#ifdef FEATURE_TIERED_COMPILATION` as `false` so the compiler should be able to remove it as dead code. What is your preference?

Comment (AaronRobinsonMSFT) [reply src/coreclr/src/vm/jitinterface.cpp]: Oh but the flag clearing isn't.... boo. Okay. I will remove it.

**Pull request merged**